File: server/python-thesis-project/api/transactions/delivery_transaction.py
import random

from django.utils import timezone

import api.utils as utils
from api.models import Customer, NewOrder, OrderLine, Orders


def do_delivery():
    # Get Warehouse
    warehouse = utils.get_random_warehouse()
    w_id = warehouse.w_id

    # Get 10 District
    for d_id in range(1, 11):
        # Get Min New Order
        new_order = NewOrder.objects.filter(
            no_w_id=w_id, no_d_id=d_id).order_by('no_o_id')
        if len(new_order) < 1:
            continue

        new_order = new_order[0]
        # Get Order
        order = Orders.objects.get(
            o_d_id=d_id, o_w_id=w_id, o_id=new_order.no_o_id)

        # Sum Order Line amount
        order_lines = OrderLine.objects.filter(
            ol_o_id=order.o_id, ol_d_id=d_id, ol_w_id=w_id)
        order_line_amount_sum = 0
        for order_line in order_lines:
            order_line_amount_sum += order_line.ol_amount

            # Update Order Line
            order_line.ol_delivery_d = timezone.now()
            order_line.save(update_fields=['ol_delivery_d'])

        # Delete New Order
        new_order.delete()

        # Update Order
        carrier_id = random.randint(1, 10)
        order.o_carrier_id = carrier_id
        order.save(update_fields=['o_carrier_id'])

        # Get Customer
        customer = Customer.objects.get(c_id=order.o_c_id, c_d_id=d_id, c_w_id=w_id)

        # Update Customer
        customer.c_balance += order_line_amount_sum
        customer.save(update_fields=['c_balance'])

File: server/js-thesis-project/src/transactions/delivery_transaction.js
const {
  get_random_warehouse,
  random_int
} = require('../utils');

async function do_delivery(models) {
  // Get Warehouse
  const warehouse = await get_random_warehouse(models)
  const w_id = warehouse.w_id

  // Get 10 District
  for (let d_id = 1; d_id < 11; d_id++) {
    // Get Min New Order
    const new_orders = await models.new_order.findAll({
      where: {
        no_w_id: w_id,
        no_d_id: d_id
      },
      order: [
        ['no_o_id', 'ASC'],
      ]
    })
    const new_order = new_orders[0]
    if (new_order == null) {
      return;
    }

    // Gert Order
    const orders = await models.orders.findAll({
      where: {
        o_d_id: d_id, o_w_id: w_id, o_id: new_order.no_o_id
      }
    })
    const order = orders[0]
    if (order == null) {
      return;
    }

    // Sum Order Line amount
    const order_lines = await models.order_line.findAll({
      where: {
        ol_o_id: order.o_id, ol_d_id: d_id, ol_w_id: w_id
      }
    })
    let order_line_sum = 0
    for (let idx = 0; idx < order_lines.length; idx++) {
      const order_line = order_lines[idx]
      order_line_sum += order_line.ol_amount

      // Update Order Line
      const now = new Date().toUTCString();
      await models.order_line.update({ol_delivery_d: now}, {
        where: {
          ol_w_id: order_line.ol_w_id,
          ol_d_id: order_line.ol_d_id,
          ol_o_id: order_line.ol_o_id,
          ol_number: order_line.ol_number,
        }
      })
    }

    // Delete New Order
    await models.new_order.destroy({
      where: {
        no_w_id: w_id,
        no_d_id: d_id,
        no_o_id: new_order.no_o_id
      }
    })

    // Update Order
    const o_carrier_id = random_int(1, 10)
    await models.orders.update({o_carrier_id}, {
      where: {
        o_w_id: order.o_w_id,
        o_d_id: order.o_d_id,
        o_id: order.o_id
      }
    })

    // Get Customer
    await models.customer.findAll({
      where: {
        c_id: order.o_c_id, c_d_id: d_id, c_w_id: w_id
      }
    })

    // Update Customer
    const c_balance = order_line_sum
    await models.customer.update({c_balance}, {
      where: {
        c_id: order.o_c_id, c_d_id: d_id, c_w_id: w_id
      }
    })
  }
}
module.exports = do_delivery;
